common/backoff/retry.go

```diff
@@ -92,29 +97,52 @@ func NewConcurrentRetrier(retryPolicy RetryPolicy) *ConcurrentRetrier {
 	return &ConcurrentRetrier{retrier: retrier}
 }
 
-// Retry function can be used to wrap any call with retry logic using the passed in policy
+// Retry function can be used to wrap any call with retry logic using the passed
+// in policy. A `nil` IsRetryable predicate will retry all errors. There is a
+// context-aware version of this function: RetryContext.
 func Retry(operation Operation, policy RetryPolicy, isRetryable IsRetryable) error {
+	ctxOp := func(context.Context) error { return operation() }
+	return RetryContext(context.Background(), ctxOp, policy, isRetryable)
+}
+
+// RetryContext is a context-aware version of Retry. Context
+// timeout/cancellation errors are never retried, regardless of IsRetryable.
+func RetryContext(
+	ctx context.Context,
+	operation OperationCtx,
+	policy RetryPolicy,
+	isRetryable IsRetryable,
+) error {
 	var err error
 	var next time.Duration
 
+	if isRetryable == nil {
+		isRetryable = func(error) bool { return true }
+	}
+
 	r := NewRetrier(policy, SystemClock)
-	for {
-		// operation completed successfully.  No need to retry.
-		if err = operation(); err == nil {
+	for ctx.Err() == nil {
+		//  completed successfully.  No need to retry.
+		if err = operation(ctx); err == nil {
 			return nil
 		}
 
 		if next = r.NextBackOff(); next == done {
 			return err
 		}
 
-		// Check if the error is retryable
-		if isRetryable != nil && !isRetryable(err) {
+		// Check if the error is retryable (err is known to be non-nil here)
+		if err == ctx.Err() || !isRetryable(err) {
```

Comment:
I'm not sure I understand the following condition:
```
err == ctx.Err()
```
Is this to handle context cancellation?  timeouts? etc

Reply:
This tests to see if the err returned by the func is due to the context ending (either due to cancellation or timeout). In such cases, we never retry, regardless of what `isRetryable` says.

Reply:
FWIW, my philosophy is that if an expert questions your code, it deserves a comment ... I'd love to see your (matt) reply inserted as a comment here.

Reply:
I did clarify the comment (GH shows the diff from the point in time when Samar asked the question, not the update). 

> 	// stop retrying if context has expired or the error
>	// is not retryable (err is known to be non-nil here)

And the function comment describes the expected behavior

> ... Context timeout/cancellation errors are never retried, regardless of IsRetryable.

So I'm hoping that covers things reasonably well.